### PR TITLE
Fixes issue 206, removed Nimbus JWT and BouncyCastle, adds Vertx JWT

### LINKED ISSUE #206: Remove JWT dependency on org.bouncycastle.bcpkix-jdk15on, replace com.nimbusds.nimbus-jose-jwt
# Removing dependency on BouncyCastle

The demo code uses PKCS#1 RSA SHA-256 certs. That works with security providers on all modern JDKs; we don't need EllipticCurve (EC) support for non-Oracle JDKs via BouncyCastle.

I successfully removed it and generated 2048 bit RSA certificate stored as PKCS#8 pem and it works on OpenJDK 11 on Linux. Will try on Windows too.

# Replacing Nimbus-jose-jwt 

I would like to replace this library with something else. The reason is that it has a hard dependency on Jackson JSON called Nimbus json-smart and poisons the MicroProfile demo app's classpath with it.

I believe MicroProfile applications should use JSONB as that is in the spec.

I successfully replaced the implementation with vertx-auth-jwt, i.e. the diff for the demo app pom is:
```
- com.nimbusds.nimbus-jose-jwt 
- org.bouncycastle.bcpkix-jdk15on
+ io.vertx.vertx-auth-jwt
```

It works with Quarkus. I will be testing it with all other servers to see if there are any problems. If there are glitches, I pick something else. The main point is not to use verx, the point is to get rid of json-smart.

# Motivation

I am working on Quarkus support and while it can work with Nimbus-jose-jwt + Nimbus json-smart in Java mode, it cannot build a native executable from it, because of what json-smart does at runtime. I believe that as long as it works for all servers, there is no reason to explicitly depend on Nimbus' package.

PR demonstrating the change comes shortly.

I would like to ask for opinions, especially @rdebusscher as he is the JWT demo author.

### PR DESCRIPTION
**Pull request by Karm**
Hello, this commit fixes #206 by replacing dependency on org.bouncycastle.bcpkix-jdk15on and com.nimbusds.nimbus-jose-jwt
with io.vertx.vertx-auth-jwt.

It also adds a script that generates expected certificates and JKS (for Liberty) and updates KumuluzEE key.

## Commit tested with these configurations:

```

curl -O -J http://localhost:8080/api/project?supportedServer=HELIDON \
&artifactId=helidonall
curl -O -J http://localhost:8080/api/project?supportedServer=HELIDON \
&selectedSpecs=JWT_AUTH&selectedSpecs=REST_CLIENT&artifactId=helidon

curl -O -J http://localhost:8080/api/project?supportedServer=THORNTAIL_V2 \
&artifactId=thorntail_v2all
curl -O -J http://localhost:8080/api/project?supportedServer=THORNTAIL_V2 \
&selectedSpecs=JWT_AUTH&selectedSpecs=REST_CLIENT&artifactId=thorntail_v2

curl -O -J http://localhost:8080/api/project?supportedServer=LIBERTY \
&artifactId=libertyall
curl -O -J http://localhost:8080/api/project?supportedServer=LIBERTY \
&selectedSpecs=JWT_AUTH&selectedSpecs=REST_CLIENT&artifactId=liberty

curl -O -J http://localhost:8080/api/project?supportedServer=KUMULUZEE \
&artifactId=kumuluzeeall
curl -O -J http://localhost:8080/api/project?supportedServer=KUMULUZEE \
&selectedSpecs=JWT_AUTH&selectedSpecs=REST_CLIENT&artifactId=kumuluzee

curl -O -J http://localhost:8080/api/project?supportedServer=PAYARA_MICRO \
&artifactId=payara_microall
curl -O -J http://localhost:8080/api/project?supportedServer=PAYARA_MICRO \
&selectedSpecs=JWT_AUTH&selectedSpecs=REST_CLIENT&artifactId=payara_micro

curl -O -J http://localhost:8080/api/project?supportedServer=TOMEE \
&artifactId=tomeeall
curl -O -J http://localhost:8080/api/project?supportedServer=TOMEE \
&selectedSpecs=JWT_AUTH&selectedSpecs=REST_CLIENT&artifactId=tomee
```

### CODE REVIEW

File: src/main/resources/files/MPJWTToken.java.tpl [hunk 1]
```diff
@@ -1,7 +1,7 @@
 package [# th:text="${java_package}"/].secure;
 
-import net.minidev.json.JSONArray;
-import net.minidev.json.JSONObject;
+import io.vertx.core.json.JsonArray;
```
Comment (Emily-Jiang) [inline src/main/resources/files/MPJWTToken.java.tpl]: Why don't you use Javax.json.JsonArry and JsonObject?

**Comment by Karm** [comment]
Hello @Emily-Jiang 
> Is there equivalent functionality in Keycloak? It is a bit odd to depend on vert.x for JWT.

We need a JWT library, unless we want to include a lot of somewhat irrelevant boilerplate code into the examples source base.

## .JSONArray / Object
We used to depend on [Nimbus](https://mvnrepository.com/artifact/com.nimbusds/nimbus-jose-jwt) and we haven't been using Javax.json.JsonArry and JsonObject either, we were using net.minidev.json.JSONArray etc.  vert.x jwt, the same as Nimbus jwt, has its own JsonObject and JsonArray. Unlike Nimbus that pulls in Jackson, vert.x operates with JSON B and that is more in line with MicroProfile.

## vert.x jwt
This pull request swaps one implementation for another, adjusting code to adhere to the another library's  API. We are not pulling the whole vert.x, just the necessary minimum.
We are also getting rid of BouncyCastle as it is not needed (we don't use Elliptic Curve crypto in the example). It was a dependency of Nimbus.

## Keycloak
It has nothing to do with Keycloak or I don't understand the question. Could you rephrase, please?

## Replacing Nimbus jwt
I tested the example using vert.x jwt works with TOMEE, PAYARA_MICRO, LIBERTY, THORNTAIL_V2, so I deem it a fit one-for-one replacement for Nimbus.

The motivation, as explained on #206, is to make Quarkus native build easily possible for MicroProfile Starter JWT_AUTH example. I believe it is O.K. as long as it works for all servers.

**Comment by rdebusscher** [comment]
Or we can use Jose4j (https://bitbucket.org/b_c/jose4j/wiki/Home) to keep it vendor neutral.

**Comment by Emily-Jiang** [comment]
ok @Karm ! I took at a look at the vert.x a bit more. It is ok to use its JWT util to generate JWT. The reason I brought up Keycloak as it is well-known security tool and thought to ask you to check whether it offers the equivalent functionality for JWT generation.

**Comment by Karm** [comment]
Not merging to 0.9.4 as it lacks 1 more ack.    

We can cut 0.9.4 later   (@rdebusscher )

File: src/main/resources/files/liberty/TestSecureController.java.tpl [hunk 2]
```diff
@@ -75,28 +67,26 @@ public class TestSecureController {
 
         token.setGroups(Arrays.asList("user", "protected"));
 
-        JWSObject jwsObject = new JWSObject(header, new Payload(token.toJSONString()));
-
-        // Apply the Signing protection
-        JWSSigner signer = new RSASSASigner(key);
+        return provider.generateToken(new JsonObject().mergeIn(token.toJSONString()), new JWTOptions().setAlgorithm("RS256"));
```
Comment (rdebusscher) [inline src/main/resources/files/liberty/TestSecureController.java.tpl]: Probably something from VertX JWT library, but why do you need to define RS256 here again? We did it already when creating the 'provider'.

**Comment by Karm** [comment]
Hold on merging, I need to resolve Kumuluzee - probably a small conf error.

File: src/main/resources/files/liberty/TestSecureController.java.tpl (hunk 2 above)
Comment (Karm) [reply src/main/resources/files/liberty/TestSecureController.java.tpl]: @rdebusscher It has to be like that. It is probably Vertx JWT UX thing. If I remove it, I get 
```
java.lang.RuntimeException: Algorithm not supported
    at io.vertx.ext.jwt.JWT.sign(JWT.java:347)
    at io.vertx.ext.auth.jwt.impl.JWTAuthProviderImpl.generateToken(JWTAuthProviderImpl.java:166)
```

So the current code as committed it correct.

**Comment by Karm** [comment]
> I would move _gencerts.jks_ out of _src/main/resources_ as it is now included in the war on the site (not a big issue but unnecessary)
> Can you document the usage and basic stuff around _gencerts.jks_ in the technical documentation of the readme.
> 
> Looks good for the rest

@rdebusscher Addressed

**Comment by Karm** [comment]
@rdebusscher I would like to have this merged. Could you re-review, please?

 * I tested it with all servers; configs marked in the description of the PR
 * CQ is not needed as the dependency on Vert.x is in generated code
 * Questions addressed

**Comment by Karm** [comment]
We have 2 acks, I am merging this change.
